independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/RunnerClassLoader.java

```diff
@@ -76,29 +72,27 @@
                 //fall through
             }
         }
-        synchronized (getClassLoadingLock(name)) {
```

Comment:
Synchronizing here doesn't seem necessary as the only update is performed in `accessingResource` which does its own synchronization

Reply:
I agree, this synchronization is not necessary, however if you remove it, you must wrap `defineClass` in a try/catch, because it becomes implicitly racy.  Something like this:

```java
        try {
            return defineClass(name, data, 0, data.length, resource.getProtectionDomain());
        } catch (LinkageError e) {
            final Class<?> loadedClass = findLoadedClass(name);
            if (loadedClass != null) {
                return loadedClass;
            }
            throw e;
        }
```

Reply:
Wow, I never knew this!

I don't see anything in the javadoc that mentions this. Do you have any resources about understanding why this is necessary?

Reply:
Yes, basically it's a straightforward race.  Some thread "A" can come along trying to load the same class, so that the first `findLoadedClass(name)` on 75 returns `false`.  Then while "A" loads the byte array, another thread "B" races in and loads the same class, so when thread "A" gets to `defineClass`, it encounters a duplicate class definition error.

This problem was first observed in JBoss Modules about 10 years ago.

Reply:
Gotcha, thanks!